LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsData.cs

```diff
@@ -40,5 +34,106 @@ public static class LnsData
+                              JsonReader.Property("DevAddr", JsonReader.UInt32()),
+                              JsonReader.Property("FCtrl", JsonReader.Byte()),
+                              JsonReader.Property("FCnt", JsonReader.UInt16()),
+                              JsonReader.Property("FOpts", JsonReader.String()),
+                              JsonReader.Property("FPort", JsonReader.Byte()),
+                              JsonReader.Property("FRMPayload", JsonReader.String()),
+                              JsonReader.Property("MIC", JsonReader.Int32()),
+                              RadioMetadataProperties.DataRate,
+                              RadioMetadataProperties.Freq,
+                              RadioMetadataProperties.UpInfo,
+                              (_, MHdr, DevAddr, FCtrl, FCnt, FOpts, FPort, FRMPayload, MIC, dr, freq, upInfo) =>
+                                new UpstreamDataFrame(new MacHeader(MHdr), new DevAddr(DevAddr), new FrameControl(FCtrl), FCnt, FOpts, new FramePort(FPort), FRMPayload, new Mic((uint)MIC),
+                                                      new RadioMetadata(dr, freq, upInfo)));
+        /*
+         * {
+              "msgtype" : "jreq",
+              "MHdr"    : UINT,
+              "JoinEui" : EUI,
+              "DevEui"  : EUI,
+              "DevNonce": UINT,
+              "MIC"     : INT32,
+              ..
+              RADIOMETADATA
+            }
+         */
+
+        private static IJsonProperty<T> EuiProperty<T>(string name, Func<ulong, T> factory, char separator = '-') =>
+            JsonReader.Property(name,
+                                from s in JsonReader.String()
+                                select Hexadecimal.TryParse(s, out var eui, separator)
+                                     ? factory(eui)
+                                     : throw new JsonException($"Could not parse {name} as {typeof(T)}."));
+
+        internal static readonly IJsonReader<JoinRequestFrame> JoinRequestFrameReader =
+            JsonReader.Object(MessageTypeProperty(LnsMessageType.JoinRequest),
+                              JsonReader.Property("MHdr", JsonReader.Byte()),
+                              EuiProperty("JoinEui", eui => new JoinEui(eui)),
+                              EuiProperty("DevEui", eui => new DevEui(eui)),
+                              JsonReader.Property("DevNonce", JsonReader.UInt16()),
+                              JsonReader.Property("MIC", JsonReader.Int32()),
+                              RadioMetadataProperties.DataRate,
+                              RadioMetadataProperties.Freq,
+                              RadioMetadataProperties.UpInfo,
+                              (_, MHdr, JoinEuiLong, DevEuiLong, DevNonce, MIC, dr, freq, upInfo) =>
+                                new JoinRequestFrame(new MacHeader(MHdr), JoinEuiLong, DevEuiLong, new DevNonce(DevNonce), new Mic((uint)MIC),
```

Comment:
Please use camel casing for arguments per .NET conventions:

```suggestion
                              (_, mhdr, joinEui, devEui, devNonce, mic, dr, freq, upInfo) =>
                                new JoinRequestFrame(new MacHeader(mhdr), joinEui, devEui, new DevNonce(devNonce), new Mic((uint)mic),
```